### DISCUSSION TITLE
When using RTH with EXTRA, airplane is returning home at negative altitude

### DISCUSSION THREAD

**Discussion by jlindrud** [iNavFlight/inav#7930] [General]
## Current Behavior

I have RTH mode set to EXTRA, with RTH altitude set to 40m. So when RTH is initiated, the airplane immediately climbs 40m and then heads home. However, if RTH is initiated at an altitude of e.g. -100m (which is common at some spots where I fly), the airplane climbs to -60m, and then flies home level, at -60m. This obviously guarantees a crash, unless I take over.

## Steps to Reproduce
1. Set RTH mode to EXTRA and RTH altitude to 40m
2. Initiate RTH at -100m.
3. Aircraft climbs from -100m to -60m, and then continues flying home at -60m.

## Expected behavior
After initial climb, if aircraft is below RTH home altitude, aircraft should keep climbing until it reaches RTH home altitude.

---

- FC Board name and vendor:
Matek F411 Wing

- INAV version string:
INAV/MATEKF411_S5S6_SS2 4.0.0 Dec 20 2021 / 14:04:20 (GITDIR-N)

**Comment by stronnag** [comment]
This is documented / expected behaviour. 
EXTRA is an additive to the the current altitude. You need to set it appropriately for your site.
See also https://github.com/iNavFlight/inav/wiki/Navigation-Mode:-Return-to-Home

**Comment by jlindrud** [comment]
The purpose of RTH is to bring the aircraft home though... The described behavior is 100% guaranteed to crash the aircraft. When would it ever make sense?

You can't specify a negative RTH altitude in Inav (for obvious reasons). If an aircraft is in RTH and finds itself below the RTH home altitude, I would think it obviously needs to climb, regardless of anything else.

**Comment by MrD-RC** [comment]
I agree with @stronnag. That is how extra is intended to work. You should chose a different RTH method that is more suitable to your location, such as AT_LEAST. 

EXTRA would make sense if you are always going to be flying above your launch point, with no large obstacles; when current altitude + extra would clear any local obstacles.

**Comment by MrD-RC** [comment]
Converted to a discussion in case we want to look at the prospect of adding another RTH Mode. EXTRA_MIN_HOME_ALT or something like that. This would add extra to home alt, or climb to home alt, if below it. That can be discussed if this it to be added.

The current EXTRA RTH mode works as expected and described, so should not be changed.

**Comment by jlindrud** [comment]
Yep, I understand how Extra mode works, and agree this is actually not related to Extra as such. It is more about what happens after the Extra climb has been performed, and the aircraft finds itself in level RTH flight at a negative altitude. 

In that situation I would expect the default behavior would be to increase the target altitude to RTH home altitude.

IMO, Inav should just ensure that no matter what RTH mode is used, the aircraft doesn't try to fly home at a negative altitude.

**Reply by MrD-RC** [reply]
But if you understand how extra works, you must understand that it is working correctly. If you are flying with a negative altitude, you have chosen the wrong RTH mode for your flying environment.

**Reply by jlindrud** [reply]
Just to be clear, I understand that Extra is working the way it is documented to work, and am only suggesting how it could be improved to also handle flights at negative altitude. I use Extra all the time and think it's a great option.

My flying environment is pretty varied in that a single flight can go from mountain ridges at 800m, down to rivers at say -100m. No single RTH mode will be optimal, but Extra covers most of the bases.

Having the FC put the aircraft into level RTH flight at -50m just seems like an unforced error to me. If a new RTH mode is needed that's OK, I just thought the behavior could be baked into Extra, as I can't think of a situation where one would want an airplane returning home at negative altitude.

Hope this doesn't come across as complaining BTW :) I'm a huge fan of Inav and all the work you guys put into it.

**Reply by MrD-RC** [reply]
To be honest, that sounds like a pretty cool place to fly. 

I’m quite happy to make changes. It’s just deciding what would be the best action. This would only really effect current and extra. Assuming that if the current altitude is negative, you would want to climb too. 

It would be good to get some feedback from the other devs to get their ideas. The way I see it there are a couple of options. 

1. Leave the modes (current and extra) as they are.
2. If RTH Home Altitude is set, make this the minimum RTH Altitude. 
3. EXTRA only: If no RTH Home Altitude, make 0 + RTH Altitude the minimum.
4. Add new _MIN_HOME_ALT modes for current and extra. So that the current modes retain their functionality. 

Thinking about it, a combination of 2 and 3 are probably the best option. As you say, who wants to RTH at a negative altitude. There’s nothing to say the current definitions have to stay the same.

@stronnag @DzikuVx what are your thoughts?

**Comment by DzikuVx** [comment]
Please do not change current modes. They work like intended

niedz., 27 mar 2022, 13:20 użytkownik Darren Lines ***@***.***>
napisał:

**Reply by stronnag** [reply]
> Please do not change current modes. They work like intended

Concur, there is already plenty of choice, possibly a confusing level of choice. At least it's well documented in the wiki.
In this particular case, either `FIXED` or `AT_LEAST` would have been more appropriate choices.

**Reply by MrD-RC** [reply]
True. But the question is, should they work like that. I am coming around the the point that a negative RTH altitude is never desirable. It will always result in an impact.

**Reply by breadoven** [reply]
Never really understood the point of `EXTRA `and `CURRENT `for precisely this reason. I guess they're OK if you're flying somewhere flat although in that case `AT_LEAST `would work just as well especially if you're concerned about having enough ground clearance ?. Are `CURRENT `and `EXTRA ` worthwhile modes to have at all ?

**Reply by OptimusTi** [reply]
> True. But the question is, should they work like that. I am coming around the the point that a negative RTH altitude is never desirable. It will always result in an impact.

I'd look at what Vector did/does on RTH. There's a Home RTH/loiter Altitude (defined by altitude and radius), a "Fly Home" altitude (aircraft climbs to this altitude on RTH until it reaches Home RTH altitude) and a "Mountain Mode" altitude (the same as EXTRA but only adds to the "Fly Home" altitude. Section 7.4 of Vector manual.  Looks like INAV has most of these elements but they don't mix well.

**Reply by Jetrell** [reply]
>  is `EXTRA ` worthwhile modes to have at all ?

I fly in mountainous area's.
 When many hundreds of meters above launch altitude, if by chance I fly behind a hill and loose radio link. `EXTRA` is the only mode that ensures the model will climb above the current flight altitude, to make it over the obstruction... Otherwise it would try to fly home through that hill. 
But like all missions.. The user must be aware of their location, and set the appropriate `RTH mode` and `RTH altitude` to suit. 
I set the `EXTRA RTH altitude` as high as 200m on some flights. 
These settings can be changed in the OSD stick menu, Having an easy way to change them on the fly would be good.

**Reply by jlindrud** [reply]
Agree, having RTH altitude as an adjustable setting would be great.

**Comment by breadoven** [comment]
I use MAX if flying in hilly areas ... on the basis you need line of sight which probably means the model is always higher than the hills between you and it in order for that to be true. Or just set `RTH Altitude` a bit higher than the highest hill in the area you're flying in. You can always override the RTH climb if it's not required ... if/once you get the Rx signal back. Flying in hilly areas is always going to be a problem where RTH is concerned, there's probably no best solution other than setting an altitude that's much higher than required although you would probably breach the 120m limit in that case.

**Reply by Jetrell** [reply]
`MAX` works most of the time.
The situation I've encountered, was if I was flying well above launch, and flying through a valley between two mountains. Then without being aware of it. The valley lead me to fly beyond LOS, and behind one of those mountains, and it block the RC link. 
The only way out, is to use `EXTRA` set to the correct climb altitude, to make it over that mountain.

An excellent feature in this case. Would be for the Nav system to be plotting Waypoints as the model travels(e.g every 80meters).
 But only keep them for the previous 500meters.
This way the model could back track (including altitude) if a FS occurred; until it gets the RC link again.

**Comment by breadoven** [comment]
I have thought RTH "Trackback" would be useful. Really nothing more than the Mission Planner that already exists set up to log WPs automatically and then switch to WP mission on Failsafe flying the WPs in reverse. And as you say it wouldn't need that many WPs if it's just to fly back to regain the signal.

**Reply by Jetrell** [reply]
@breadoven The amount of crashes I've read of over the years on forums, caused by RTH setting that didn't suit the topographical location, is too many to remember.
 A short range "Trackback" feature would have gotten most of them out of trouble.

It would just have to be determined, whether full straight line RTH takes over after the Trackback finishes, if there is major signal loss. And if the WP logging intervals and Trackback distance should be user configurable. (within reason)

**Comment by stronnag** [comment]
See also the `agh_smart_rth` branch (from 2018-04). 

Some notes from a [somewhat sub-optimal test session](https://github.com/stronnag/mwptools/wiki/Testing-Smart-RTH)

**Comment by sensei-hacker** [comment]
Agreed with Mr D's 2 and 3.

It will never work to RTH at al altitude lower than the home altitude, unless you're in a submarine and launched from a boat.

Any time home is on the ground, negative total RTH altitude will always crash. So while it technically marches documentation, it guarantees a crash and is therefore not sensible.
